config/database.yml

```diff
@@ -1,24 +1,23 @@
+<% name = Rails.application.class.parent.to_s.underscore %>
+
 default: &default
   adapter: postgresql
-  encoding: unicode
```

Comment:
Hocam encoding belirtmeye gerek yok mu? Fail eden test onu yakalamış.

Reply:
Tam yazıyordum onu :-)  Aşağıda yazdım.

Reply:
Aslında şart koşmak gibi bir niyetim yoktu, Vagrant ile çalıştığımdan göremedim :-P  Ne dersin böyle  mi devam edelim, yoksa default pg user/pass ile devam mı edelim?  Bunlar lokal ortamlar, ayrı user/pass'in faydası faydası ne olabilir?  Çok çok kazayla geliştirme ortamlarında hasarı engellemiş oluruz.  Bu pratiği de bir inceleyeyim, büyükler ne diyor bir bakayım.

Reply:
Postgres'te bildiğim kadarıyla doğru pratik her bir veritabanı için ayrı bir kullanıcının olması ve o kullanıcının yalnızca ilgili veritabanına yetkili olması. Genelde geliştiricilerin yerelde kullandığı pratikler ise:

- Postgres kullanıcısını abuse edip tüm veritabanlarında kullanmak.
- Postgres kullanıcısını tüm yetkilendirmelerden muaf tutup direk şifresiz şekilde veritabanlarına erişmek.
- MySQL kullanıldığında ise root kullanıcısını abuse edip tüm veritabanlarını onla kullanmak.

Tabii ki bunların hepsi yanlış işler. Bu pratiklerin tümü geliştirici makinasını da saldırılara epey bir açık hale getiriyor. Bence bu PR'de gelen şekliyle devam etmek doğru pratik ve çok daha güvenli.

Reply:
Ben yıllardır böyle yapıyorum: "user/pass per app".  Şimdi okuyorum, dediğin gibi yaygın ve doğru pratik bu.  O halde bir test ekleyelim.  (Rails yapılandırmalarını yüklendikten sonra inspect eden bir metod olacaktı, ona bakayım)

Reply:
a ha 
![image](https://user-images.githubusercontent.com/47842/44958425-d260b900-aee8-11e8-8358-88792fb7f150.png)

Reply:
Encoding konusunu inceledim.

`en_US.UTF-8` etkin sunucuda pg encoding'i nasıl görüyor?

```sh
$ sudo -u postgres psql --dbname=nokul_development -c "SHOW client_encoding;"
 client_encoding
-----------------
 UTF8
(1 row)
```
Davranış şu (dokümante etmişler, tembellikten alıntılamıyorum):  Postgres çalıştığı ortamdaki yerele bakarak UTF-8 ayarlıyor.

Bu durumda encoding'in tek yararı OS X (default kodlama Western sanırım) ve Windows (en son UTF-16 biliyorum) ortamlarında geliştirme yaparken (development ve test) sorun olmasını önlemek.

Peki önceki iletide söylediğim gibi url tanımlandığında encoding ayarı eziliyor mu gerçekten.  Onu da denedim.  Yanlış tespit yapmışım.  Evet `DATABASE_URL`'den etkilenmemek için url'li tanım yapmalıyız ama `default` sekmesinde tanımlanan encoding ayarı aynen geçerli oluyor.  Rails şöyle davranıyor.  Url yoluyla veritabanı ayarları verildiğinde eksik bir seçenek default'tan geliyorsa kabul ediyor.  Şöyle:

```
default
  encoding: unicode
development:
  <<: *default
  url: postgresql://foo:bar@localhost/db?encoding=western
```

yapılandırmasında development için encoding query sentaksıyla girildiğinden kodlama western oluyor.  Ama bu girilmemişse encoding default'tan kabul ediliyor ve unicode oluyor.

Özetle, `unicode` default'ta tanımlanması tamamen redundant değil.  Mac OS X vb UTF-8 default olmayan development ve test ortamlarında gerekiyor (bu sistemlerde default kodlama UTF-8 yapılmamışsa).  Production vb için ise gerekmiyor, production konuşlandırmayı daima UTF-8 Linux ortamlarında yaptığımızdan.

Karar anı :-P  Ayarı ekliyorum.